### ISSUE TITLE
Finding secret and tokens (Android only | Requires adb) - script error

### ISSUE THREAD

**Issue by kalatos86** [jonkristian/alko#2]
Hello,

I am trying to go through the below procedure described in the documentation:

Manual or via HACS

If you're using HACS you can add this repo as a custom repository and install, otherwise download or clone and copy the folder custom_components/alko into your custom_components/. Be sure to restart.
Finding secret and tokens (Android only | Requires adb)

The AL-KO API is unfortunately not open for 3rd party applications. In order for this component to work you will have to have set up the AL-KO inTOUCH app and a [working adb connection to your phone](https://developer.android.com/studio/command-line/adb). To extract your secret and tokens I've created a very simple and crude script that will output these in the terminal.

    Download [intouch-credentials.sh](https://raw.githubusercontent.com/jonkristian/alko/master/tools/intouch-credentials.sh) (located in tools folder).
    Then chmod +x intouch-credentials.sh and ./intouch-credentials.sh to run.
    Now unlock your phone and click confirm the backup operation (do not encrypt).
    This should print your client_secret, access_token and refresh_token the latter two is required during the integration setup step.
    Add the following to your configuration.yaml.

alko:
  client_id: inTouchApp
  client_secret: your_client_secret

After a restart you should now be able to add the integration via the integrations page, remember that tokens are short-lived 30 minutes.

Unfortunately while trying to run the script I am getting error message that such file or directory does not exist:

![obraz](https://user-images.githubusercontent.com/75856476/171002875-63153b79-1243-4a72-92c0-c72ea701cc87.png)

Could you be so kind and advise how to get secret and token to make the integration running on HA?

**Comment by jonkristian** [comment]
Unfortunately it seems that Alko has changed their app, so this no longer works as intended and I've not yet been able to figure out a workaround, so I will keep this open.

Anyone else wants to take a stab at it, that would be great!

**Comment by aspic** [comment]
Hmm, would really like to get this integration going. Has anyone been able to figure out auth details?

**Comment by aspic** [comment]
My guess is that AL-KO in their latest app update disabled backups in AndroidManifest.xml (`android:allowBackup="false")`. Making it a bit more cumbersome to extract these credentials.

**Comment by jeppevinkel** [comment]
Possibly yeah. I've attempted to see what info I could gather by listening to https traffic on my phone, but without success as the AL-KO app doesn't seem to be proxy aware and I haven't rooted my phone so I can't force it.

**Comment by aspic** [comment]
A way to verify this / test would be to enable backup again; https://stackpointer.io/mobile/android-enable-adb-backup-for-any-app/462/

But this is a rather cumbersome process to get the integration working.

**Comment by patrick-zinner** [comment]
It's a pitty that vendors like AL-KO don't provide an API for their community. I hoped that I can use this integration. I can only confirm what other people here wrote. 

Anyway, it would be nice if we would all write an email to them and ask them for this.

**Comment by jeppevinkel** [comment]
I did write an email, but I don't think I ever got a response.

**Comment by jonkristian** [comment]
I did get a response, not much but it's something. 

October 21.

Dear Jon,

We rolled out a new major version of our app on a whole new tech stack which obviously does not allow the adb backup feature any more. We actually did not remove this on purpose but were just not aware that this is actually a feature used by customers.
 
As said, opening our app to third party developers is planed for the near future and we hope with this, we can help out the Home Assistant community with integrating their Robolinho devices with their Home Assistant controlled smart homes.

Best regards,
Your AL-KO Gardentech team

So I guess we will just have to wait.

**Comment by MiG-41** [comment]
I'm thinkig about Robolinho... And it is only a case of getting client_id and client_secret from app ,and i you have i already then integration works ? 
If yes , is it possible to install older version  of app and get this data ?

**Comment by szerwi** [comment]
Hi,
Is there any update regarding this issue? Is it currently possible to get the neccessary client id and secret?
Is there a way to use older app for the keys extraction?

I've heard about Al-ko integration with Home Assistant via IFTTT, but seems like it needs HA cloud subscription in order to work.

**Comment by jeppevinkel** [comment]
I'm afraid there's no updates so far. You can use IFTTT with HA without a cloud subscription. You just need to have your HA instance acessible over the web to receive events, but you can trigger events just fine either way. You can see the relevant info here https://www.home-assistant.io/integrations/ifttt/.

One issue I have with the IFTTT integration is that it doesn't have all features. Like you can't use IFTTT to start the mower from a specific start point. It will always start right at the base. That's not useful for me due to the complex shape of the lawn in question.

**Comment by szerwi** [comment]
@jeppevinkel Ok, thanks for the information.
Has anyone tried to get the keys using older app? As I understand, the API still works and the problem is just with getting the required ID and secret? Or did AL-KO completely disabled their API so even the users that got the ID/secret earlier cannot use the integration?

**Comment by jonkristian** [comment]
I haven't tried, could be interesting to test. I don't think they've locked us out, as they stated it was not their intention when changing that specific feature in the app so it might work.

**Comment by szerwi** [comment]
I have downloaded InTouch 3.4.2 app from APKPure and (probably) managed to get the keys.
Now I get another error during the integration, but I've created another issue for that.
However, there is a change that it will work after I resolve that issue

**Comment by jeppevinkel** [comment]
Oh that's great. I'll give that a try too when it becomes season for the mower again.

**Comment by MiG-41** [comment]
Finaly bought Robolinho 2000W ,and want to integrate.  Red about it earlier ,and thought i was prepared , but no succes...

1. Script not showing me client_secret  .... It was only once on several tries .  With access_token and refresh_token there is no problem. So my question is , is client_secret one for my account , or it is different every new login ? My it have 18 characters with special characters like { } : % ....
2.  line in configuration.yaml "client_id:  inTouchApp" can stay , i don't have to replace word "inTouchApp" with another key ?
3.  access_token has app 1900 characters ? Is that ok ?
4.  finally , when i put all this i receive an error in log:

Error setting up entry Alko for alko
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 387, in async_setup
    result = await component.async_setup_entry(hass, self)
  File "/config/custom_components/alko/__init__.py", line 120, in async_setup_entry
    hass.config_entries.async_setup_platforms(entry, PLATFORMS)
AttributeError: 'ConfigEntries' object has no attribute 'async_setup_platforms'

Maybe fom this there is a info what is wrong ?

**Comment by MiG-41** [comment]
Seems , like something is going on on this data , since i use them in HA i can't login anymore on my android phone.....
o question is , that  this "async_setup" is something related with new update of HA  ,and this is the problem with starting integration ?
@jonkristian  , any chance to look on this ?

**Comment by esukram** [comment]
Hi.
Not sure if you know, that AL-KO is exposing their [OpenID endpoints](https://[idp.al-ko.com/.well-known/openid-configuration](https://idp.al-ko.com/.well-known/openid-configuration)), along with a ["debug" page](https://idp.al-ko.com/diagnostics).

This may help.

**Comment by jonkristian** [comment]
@esukram This is interesting, I'll take a look.
@MiG-41 The integration was working on earlier android apps but not the latest, as you might know it is not entirely straight forward to get it working. But I will look into the recent changes from al-ko, maybe there's hope.

**Comment by MiG-41** [comment]
@jonkristian ,yes it was initially my idea to use old apps from apps backup ant try it there. But not sure , if i have correct data... 
Main important question is , at your side your adon is still working ?

**Comment by jonkristian** [comment]
No, i don't have it working at the moment, haven't had time to look at it again since it stopped working when they updated their app, however I think someone got it working by reverting to older APK. It seems that they might actually open up the api now though, so I think I will focus on that once it happens.

**Comment by MiG-41** [comment]
@jonkristian , thanky you for answer....
I stilll have installed on my old rooted phone inTouch ver 3.4.2 , and in 9 May 2023 i was able to see device on it , and get the credentials via adb ( but since in HA had errors , i was not sure that credentials was ok ). 
Now i'm not able to see device on it , it says me "Unknown error 403 , try again..."

**Comment by jonkristian** [comment]
@MiG-41 I took a look at your error and it is indeed a deprecation, so I will take a look at the code this weekend to fix that at least. 

@esukram I've been looking at the changes they've made, but did not manage to find any documentation yet. I'll send them an email this weekend to ask about it.

**Comment by matbgn** [comment]
Woaw 😍 I'm son thankful that you will investigate, I think it's time for a coffee

**Comment by sikysikov** [comment]
@jonkristian was there any reply from the producer?

**Comment by jonkristian** [comment]
They never got back to me, I will write them again.

**Comment by jonkristian** [comment]
I haven't had time to test so I can only speculate, but given the recent changes with their api i think maybe not. But these changes seems to indicate something is happening on their end in regard to allowing access maybe 🤔, so best bet would be to get a response from them and hope that is the case. If anyone can get som more information it would be nice to collect it in here.

**Comment by Whomeyes** [comment]
Swedish support for al-ko says open api is not in their plans atm

**Comment by jonkristian** [comment]
> Swedish support for al-ko says open api is not in their plans atm

That is not the signal I've been getting, from i suppose Germany. But still no reply on my last email though, I think it would be a good idea for everyone to send along another email to try and push this in the right direction.

**Comment by matbgn** [comment]
Donne!

**Comment by esukram** [comment]
Done

**Comment by matbgn** [comment]
> Done!

Answer received today: 

Sehr geehrter Kunde / Dear customer / Gentile Cliente / Cher client,
ich habe mich bei unserer Marketing / IT Abteilung erkundigt. Dieser Punkt ist in der Umsetzungsphase. Ist aber noch nicht umgesetzt. Einen Zeithorizont konnte man mir noch nicht bekannt geben. Meine Annahme ist dass dies bis ende Jahr umgesetzt wird.

Besten Dank für Ihr Verständnis und entschuldigen Sie die negative Rückmeldung.

Kind regards
Your customer Service Team - AL-KO Gardentech Switzerland

**Comment by jonkristian** [comment]
Nice @matbgn ! Also thank you very much for your donation. Let's just hope they open it up sooner rather than later. It's been a long wait.

**Comment by jonkristian** [comment]
So I just wanted to update you guys. I too got a reply from Swedish support stating that it is not in their plans, then I forwarded earlier responses from Germany and after a couple of days I received a new reply saying it was some kind of miscommunication and that they are indeed working on it. Too bad the summer is almost over, but at least some positive news.

**Comment by sopparus** [comment]
summer is coming! any news?

**Comment by jonkristian** [comment]
Haven't heard anything yet, we can only hope, it would be a nice surprise for sure.

**Comment by matbgn** [comment]
Is their any other scientific path we can follow or wait and trash the Alko Bot away? 

I mean is there any kind of Valetudo https://github.com/Hypfer/Valetudo alternative there or other community development that you could imagine @jonkristian? I will for sure spent a few ours of my time in it, but if you have an idea I will base my future work on it.

**Comment by jonkristian** [comment]
To my knowledge I don't think there is, but it would indeed be nice if it were :-)

**Comment by kalatos86** [comment]
Back to the question is there any chance to integrate it anyhow with the HA?

**Comment by Whomeyes** [comment]
Got this mail today,
yes, this is still the plan. We are currently rolling this out. As soon as this is available, we will let you know. I expect it to be ready within the next 2 weeks.

Best regards
Your AL-KO Gardentech team

**Comment by jeppevinkel** [comment]
Sounds like great news if true.

**Comment by jonkristian** [comment]
That is indeed great news! Hopefully it will happen and we can get this project going again :)

**Comment by Lanng0110** [comment]
If you didn't get the information already, the api access has been released.
Request api access https://alko-garden.com/api-access
I haven't tested the integration yet, but from Postman the api seems to be working fine.
The access token you create has a ttl for 3600
